tests/t0118_gateway_car_test.go

```diff
@@ -13,44 +14,179 @@ func TestGatewayCar(t *testing.T) {
-					Header("X-Content-Type-Options").
-						Hint("CAR is streamed, gateway may not have the entire thing, unable to calculate total size").
-						Equals("nosniff"),
-					Header("Accept-Ranges").
-						Hint("CAR is streamed, gateway may not have the entire thing, unable to support range-requests. Partial downloads and resumes should be handled using IPLD selectors: https://github.com/ipfs/go-ipfs/issues/8769").
-						Equals("none"),
+						Contains(`attachment; filename="{{cid}}.car"`, fixture.MustGetCid()),
 				).Body(
 				IsCar().
-					HasRoot(fixture.MustGetCid("subdir", "ascii.txt")).
-					HasBlock(fixture.MustGetCid("subdir", "ascii.txt")).
+					HasRoot(fixture.MustGetCid()).
+					HasBlocks(
+						fixture.MustGetCid(),
+						fixture.MustGetCid("subdir"),
+						fixture.MustGetCid("subdir", "ascii.txt"),
+					).
 					Exactly().
 					InThatOrder(),
 			),
 		},
+		{
+			Name: "GET CAR with dag-scope=block",
+			Hint: `
+				dag-scope=block should return a CAR file with only the root block and a
+				block for each optional path component.
+			`,
+			Request: Request().
+				Path("ipfs/%s/subdir/ascii.txt", fixture.MustGetCid()).
+				Query("format", "car").
+				Query("dag-scope", "block"),
+			Response: Expect().
+				Status(200).
+				Body(
+					IsCar().
+						HasRoot(fixture.MustGetCid()).
+						HasBlocks(
+							fixture.MustGetCid(),
+							fixture.MustGetCid("subdir"),
+						).
+						Exactly().
+						InThatOrder(),
+				),
+		},
+		{
+			Name: "GET CAR with dag-scope=entity",
+			Hint: `
+				dag-scope=entity should return a CAR file with all the blocks needed to 'cat'
+				a UnixFS file at the end of the specified path, or to 'ls' a UnixFS directory
+				at the end of the specified path.
+			`,
+			Request: Request().
+				Path("ipfs/%s", fixture.MustGetCid()).
+				Query("format", "car").
+				Query("dag-scope", "entity"),
+			Response: Expect().
+				Status(200).
+				Body(
+					IsCar().
+						HasRoot(fixture.MustGetCid()).
+						HasBlocks(
+							fixture.MustGetCid(),
+						).
+						Exactly().
+						InThatOrder(),
+				),
+		},
+		{
+			Name: "GET CAR with dag-scope=all",
+			Hint: `
+				dag-scope=all should return a CAR file with the entire contiguous DAG
+				that begins at the end of the path query, after blocks required to verify path segments.
+			`,
+			Request: Request().
+				Path("ipfs/%s/subdir", fixture.MustGetCid()).
+				Query("format", "car").
+				Query("dag-scope", "all"),
+			Response: Expect().
+				Status(200).
+				Body(
+					IsCar().
+						HasRoot(fixture.MustGetCid()).
+						HasBlocks(
+							fixture.MustGetCid(),
+							fixture.MustGetCid("subdir"),
+							fixture.MustGetCid("subdir", "ascii.txt"),
+						).
+						Exactly().
+						InThatOrder(),
+				),
+		},
 	}
 
-	Run(t, tests)
+	transformedTests := standardCARTestTransforms(t, tests)
+	Run(t, transformedTests)
+}
+
+func standardCARTestTransforms(t *testing.T, tests SugarTests) SugarTests {
```

Comment:
That looks fine, I like the idea of using this decorator-ish approach,

We do call stuff like `Clone()` and `delete(map, k)` which moves away from our builder pattern and digs quite deep into the internals: could you move this into `tooling/helpers`?